add description and examples for `enum` keyword (#46)
Intelligence-AI/learnjsonschema.com: content/2020-12/validation/enum.markdown

  - vocabulary: applicator
    keyword: oneOf
---

>  The `enum` keyword specifies a validation constraint for an instance, defining a set of permissible values. The value of the `enum` keyword must be an array containing at least one element, and these elements should be unique. The validation succeeds if the value of the instance matches one of the elements in the `enum` array.

Reviewer: Reads well, but an extra space here?

```suggestion
The `enum` keyword specifies a validation constraint for an instance, defining a set of permissible values. The value of the `enum` keyword must be an array containing at least one element, and these elements should be unique. The validation succeeds if the value of the instance matches one of the elements in the `enum` array.
```

## Examples

{{<schema `Schema with string enum`>}}
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",

>   "type": "string",

Reviewer: If using `enum`, it is considered a good practices to *not* include `type`, as it is redundant. Can we remove it from the below examples too, and maybe even clarify this point?

>   "enum": ["red", "green", "blue"]

Reviewer: For consistency, can we add spaces after/before square brackets in all the examples?

```suggestion
  "enum": [ "red", "green", "blue" ]
```

Author: sure!

> _**Note:** When using  `enum`, it's recommended not to include `type` as it is redundant. While it's not mandatory to exclude `type` with `enum`, combining `enum` and `type` is considered an anti-pattern, so it's better to avoid doing so._

Reviewer: I think we can simplify this paragraph by making it a lot more direct:

```suggestion
_**Note:** Using the `type` keyword along the `enum` keyword is considered an anti-pattern, as `enum` constraints instances tighter than `type`._
```

}
{{</schema>}}

{{<instance-pass `Instance with value present in the enum is valid`>}}
"green"
{{</instance-pass>}}

{{<instance-fail `Instance with value not present in the enum is invalid`>}}
"black"
{{</instance-fail>}}

{{<schema `Schema with number enum`>}}
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "type": "number",

>   "enum": [2, 46, 100]

Reviewer: ```suggestion
  "enum": [ 2, 46, 100 ]
```

> "enum": [ 2, 46, 100 ]

Reviewer: Indentation issues here?

Author: oops! fixed.

}
{{</schema>}}

{{<instance-pass `Instance with value present in the enum is valid`>}}
45
{{</instance-pass>}}

{{<instance-fail `Instance with value not present in the enum is invalid`>}}
70
{{</instance-fail>}}

{{<instance-fail `Instance with value having different datatype is invalid`>}}
"2"
{{</instance-fail>}}

{{<schema `Schema with mixed-type enum`>}}
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",

>   "enum": [ "red", 123, true, { "foo": "bar" }, [ 1, 2 ], null ]

Reviewer: Awesome example :+1:

}
{{</schema>}}

{{<instance-pass `Instance with value present in the enum is valid`>}}
true
{{</instance-pass>}}

{{<instance-fail `Instance with value not present in the enum is invalid`>}}
{ "foo": "baz" }
{{</instance-fail>}}

> -  _Without specifying a type, you can utilize enum to accept values of various types._

Reviewer: As pointed out above, it is actually considered an anti-pattern to use `enum` + `type`. In these cases, we should not promote `type` at all
